### DISCUSSION TITLE
NMP timeout when i am using  any mcumgr command

### DISCUSSION THREAD

**Discussion by CEmir10** [zephyrproject-rtos/zephyr#46003] [Q&A]
I followed all the instructions from https://docs.zephyrproject.org/2.6.0/samples/subsys/mgmt/mcumgr/smp_svr/README.html. Every time i try to use mcumgr command  with nrf52840dk_nrf52840 I am getting nmp timeout error, even in case i msddisable and use -t <timeot.

crem@CHVL0365u:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ mcumgr --conntype serial --connstring "/dev/ttyACM0,baud=115200" echo hello

Error: NMP timeout
crem@CHVL0365u:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ hello
Hello, world!
crem@CHVL0365u:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ mcumgr --conntype serial --connstring "/dev/ttyACM0,baud=115200" image list
Error: NMP timeout
crem@CHVL0365u:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ mcumgr -t 30 --conntype serial --connstring "/dev/ttyACM0,baud=115200" image list
Error: NMP timeout
crem@CHVL0365u:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ mcumgr -t 30 -r 4 --conntype serial --connstring "/dev/ttyACM0,baud=115200" image list 
Error: NMP timeout
crem@CHVL0365u:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ mcumgr  -r 4 --conntype serial --connstring "/dev/ttyACM0,baud=115200" image list
Error: NMP timeout
crem@CHVL0365u:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ mcumgr --conntype serial --connstring "/dev/ttyACM0,baud=115200" echo hello
Error: NMP timeout
crem@CHVL0365u:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ mcumgr --conntype serial --connstring "/dev/ttyACM0,baud=9600" echo hello
Error: NMP timeout
crem@CHVL0365u:~/zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr$ mcumgr --conntype serial --connstring "/dev/ttyACM0,baud=9600" echo hello
Error: NMP timeout

**Comment by de-nordic** [comment]
What is your full build command?
Is there more than one ttyACM in a system and are you sure that the right one is in use?

**Reply by CEmir10** [reply]
west build -b nrf52840dk_nrf52840  -- -DOVERLAY_CONFIG='overlay-serial.conf;overlay-fs.conf;overlay-shell-mgmt.conf'
ttyACM0 is right one, because i tried to use it with putty. Do you have some idea what should i modify?

**Reply by de-nordic** [reply]
Try to break the connection and issue command again. There seems to be some problem with UART based transports and first command does not go through. The same problem does not appear when you build for bluetooh.

**Reply by CEmir10** [reply]
no, it did not help

**Reply by de-nordic** [reply]
I have placed it under dbg and found that: on freshly reset board when I issue mcmumgr echo command it is processed and proper response is generated, but once it is passed to transport (UART) nothing happens.
When I start quit and start new debug session, without resetting of the board, the process is the same but transport does send out the response.

**Reply by CEmir10** [reply]
Would you like to share with me your command from terminal. Thanks

**Reply by de-nordic** [reply]
`sudo ~/go/bin/mcumgr -t 1200 --conntype serial --connstring dev=/dev/ttyACM0,baud=115200 echo hello`
Ctrl+C and again works second time on a freshly reset board.

The 
`sudo ~/go/bin/mcumgr -t 1200 --conntype serial --connstring dev=/dev/ttyACM0,baud=115200 image list`
does not work however many times I try.
But
`sudo ~/go/bin/mcumgr -t 1200 --conntype serial --connstring dev=/dev/ttyACM0,baud=115200 echo hello`
and Ctrl+C  and
`sudo ~/go/bin/mcumgr -t 1200 --conntype serial --connstring dev=/dev/ttyACM0,baud=115200 image list`
works and from that point, till the reset, all commands work perfectly.

The BT transport works all the time.

**Reply by CEmir10** [reply]
So i paste your  first command and after2-3 sec i Ctrl+C after that  i paste your second command and I got same error
crem@CHVL0365u:~/go/bin$ mcumgr -t 1200 --conntype serial --connstring dev=/dev/ttyACM0,baud=115200 echo hello
^Ccrem@CHVL0365u:~/go/bin$ mcumgr -t 1200 --conntype serial --connstring dev=/dettyACM0,baud=115200 image list
Error: NMP timeout

**Reply by de-nordic** [reply]
Can you build this for BT? When you build this for BT it includes shell so you can either use shell or the BT for such configuration.
Can you than try:
 1) whether echo command works with BT?
 2) Check shell with putty to see if you can see the "shell" prompt?
 3) After making sure the shell works correctly (2), can you disconnect Putty from the shell and try UART update?

**Reply by CEmir10** [reply]
if i use command for ble i am getting Error: [hci0]: can't init hci: can't down device: operation not permitted, before this command i build and flash smp_svr samples using these commands  west build -b  nrf52840dk_nrf52840 -- -DOVERLAY_CONFIG=overlay-bt.conf
then i signed and flash image.

**Reply by de-nordic** [reply]
Have you been able to access shell via putty?

**Reply by CEmir10** [reply]
Yes, i am able to access ACM0 via putty.

**Reply by CEmir10** [reply]
Now it works for me with nrf42840, but now i switched to my board atsame70j19 and when i try to build it, it says me error __device_dts_ord_DT_CHOSEN_zephyr_uart_mcumgr_ORD undeclared

**Reply by de-nordic** [reply]
>Now it works for me with nrf42840

What was the problem?

> __device_dts_ord_DT_CHOSEN_zephyr_uart_mcumgr_ORD

Seems you do not have the `zephyr,uart-mcumgr = &<uart label>` in you DTS. Look at the `boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts`

**Reply by CEmir10** [reply]
I just had to add two more files to build overlay-serial.conf and overlay-shell-mgmt.conf and then it works fine.
I added already in device tree zephyr,uart-mcumgr = &uart0, but it does not work

**Comment by CEmir10** [comment]
If you try to build it for sam_e70_xplained you will get same error, but only if you enable serial, without enabling serial you can build it



Sent from my Galaxy



-------- Original message --------
From: Dominik Ermel ***@***.***>
Date: 6/2/22 12:11 (GMT+01:00)
To: zephyrproject-rtos/zephyr ***@***.***>
Cc: CEmir10 ***@***.***>, Author ***@***.***>
Subject: Re: [zephyrproject-rtos/zephyr] NMP timeout when i am using any mcumgr command (Discussion #46003)


Now it works for me with nrf42840

What was the problem?

__device_dts_ord_DT_CHOSEN_zephyr_uart_mcumgr_ORD

Seems you do not have the zephyr,uart-mcumgr = &<uart label> in you DTS. Look at the boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts

**Reply by de-nordic** [reply]
I have looked at the `boards/arm/sam_e70_xplained/sam_e70_xplained-common.dtsi` and it seems that only `usart1` is available (defined) as uart device, are you selecting the right device with the "chosen" thing?

**Comment by CEmir10** [comment]
On my board atsame70j19 i have only uart0, i can use it as console, but as i said if i MCU_BOOT_SERIAL=y and try to build it i jave an error



Sent from my Galaxy



-------- Original message --------
From: Dominik Ermel ***@***.***>
Date: 6/2/22 12:45 (GMT+01:00)
To: zephyrproject-rtos/zephyr ***@***.***>
Cc: CEmir10 ***@***.***>, Author ***@***.***>
Subject: Re: [zephyrproject-rtos/zephyr] NMP timeout when i am using any mcumgr command (Discussion #46003)


I have looked at the boards/arm/sam_e70_xplained/sam_e70_xplained-common.dtsi and it seems that only usart1 is available (defined) as uart device, are you selecting the right device with the "chosen" thing?

**Reply by de-nordic** [reply]
The mcuboot serial thing is for mcuboot not mcumgr library. It allows you to update application slot from mcuboot.
I think that we are mixing problems here.

**Reply by CEmir10** [reply]
yes i know. this is a part of device tree of my custom board / {
	aliases {
		i2c-0 = &twihs0;
		i2c-1 = &twihs1;
		i2c-2 = &twihs2;
		led0 = &green_led;
		sw0 = &sw0_user_button;
		can0 = &can_silent;
	};

	chosen {
		zephyr,console = &uart0;
		zephyr,shell-uart = &uart0;
		zephyr,uart-mcumgr = &uart0;
		zephyr,sram = &sram0;
		zephyr,flash = &flash0;
		zephyr,code-partition = &slot0_partition;
		zephyr,canbus = &can0;
	};
Whatever other board i try to build i get same error. Only Nrf boards work.
